### PR TITLE
[feat] Add transplantation augmentation

### PR DESCRIPTION
**Pull request by JanSellner**
#### Changes
This PR adds the [transplantation augmentation](https://arxiv.org/abs/2303.10972) which copies object instances between images in a batch.
image

The paper was accepted at [MICCAI2023](https://conferences.miccai.org/2023/en/) and even though originally targeted at (hyperspectral) medical images, I think it could be of general interest to the community and hence it would be cool to see this in kornia :-) 

#### Type of change

- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

#### TODO
- [x] If everything is fine, it would be great if someone could upload the image [https://files.jansellner.net/simba_mask.png](https://files.jansellner.net/simba_mask.png) which I used in the [generate_examples.py](docs/generate_examples.py) script to kornia/data, then I can change the link.

### CODE REVIEW

File: docs/generate_examples.py [hunk 1]
```diff
@@ -102,6 +102,7 @@ def main():
     BASE_IMAGEOUTDOOR_URL8: str = (
         "https://github.com/kornia/data/raw/main/kornia_banner_pixie.png"  # Response functions
     )
+    MASK_IMAGE_URL2: str = "https://files.jansellner.net/simba_mask.png"
```
Comment (JanSellner) [inline docs/generate_examples.py:105]: TODO: change link

File: docs/generate_examples.py [hunk 2]
```diff
@@ -214,6 +219,27 @@ def main():
...
+        aug = cls(*args, p=1.0)
+        # set seed
+        torch.manual_seed(seed)
+        # apply the augmentation to the image and concat
+        img_aug, _ = aug(img_in, mask_in)
+
+        output = torch.cat([img_in[0], img_in[1], img_aug[0]], dim=-1)
+        # save the output image
+        out_np = K.utils.tensor_to_image((output * 255.0).byte())
+        cv2.imwrite(str(OUTPUT_PATH / f"{aug_name}.png"), out_np)
```
Comment (edgarriba) [inline docs/generate_examples.py:239]: you can use `kornia.io.write_image` here: https://github.com/kornia/kornia/blob/master/kornia/io/io.py#L119

File: kornia/augmentation/_2d/mix/transplantation.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                [[0, 2, 2, 0, 0],
+                 [0, 0, 0, 0, 1],
+                 [0, 0, 0, 2, 0],
+                 [0, 0, 1, 0, 1],
+                 [1, 1, 1, 2, 0]]])
+    """
+
+    def __init__(
+        self,
+        excluded_labels: Optional[Sequence[int]] = None,
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/transplantation.py:122]:
```suggestion
        excluded_labels: Sequence[int] | None = None,
```
can you use typing annotations from `__future__` ?

File: kornia/augmentation/_2d/mix/transplantation.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        p: float = 0.5,
+        p_batch: float = 1.0,
+        data_keys: Optional[List[Union[str, int, DataKey]]] = None,
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/transplantation.py:125]:
```suggestion
        data_keys: list[str | int | DataKey] | None = None,
```

File: kornia/augmentation/_2d/mix/transplantation.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    ) -> Union[Tuple[Tensor, Tensor], List[Tensor]]:
+        keys: List[DataKey]
+        if data_keys is None:
+            keys = self.data_keys
+        else:
+            keys = [DataKey.get(inp) for inp in data_keys]
+
+        inputs: List[Tensor] = [image, mask, *additional_inputs]
+        if len(keys) != len(inputs):
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/transplantation.py:170]: i'd use the `KORNIA_CHECK` api here

**Review by edgarriba — COMMENTED** [review]
good one!

File: docs/generate_examples.py (hunk 1 above)
Comment (johnnv1) [reply docs/generate_examples.py:105]: You can open a PR on https://github.com/kornia/data/

File: docs/generate_examples.py (hunk 2 above)
Comment (johnnv1) [reply docs/generate_examples.py:239]: @edgarriba you know if we have some issues tracking the rewrite of this `generate examples` (aside the #2443)?

this script is worth a total rewrite, using some new functions such as `kornia.io.write_image` and simplifying as said in #2443

File: kornia/augmentation/_2d/mix/transplantation.py (hunk 5 above)
Comment (johnnv1) [reply kornia/augmentation/_2d/mix/transplantation.py:170]: +1 for others check with raise

File: test/augmentation/test_augmentation_mix.py [hunk 6]
```diff
@@ -387,3 +387,113 @@ def test_p(self, p, same_on_batch, device, dtype):
...
+            mask[i, selection] = i + 1
+
+        image_copy = image.clone()
+        mask_copy = mask.clone()
+        mask_additional_copy = mask_additional.clone()
+
+        f = RandomTransplantation(p=1, excluded_labels=[0])
+        image_out, mask_out, mask_additional_out = f(image, mask, mask_additional, data_keys=["input", "mask", "mask"])
+
+        assert torch.allclose(image, image_copy)
```
Comment (johnnv1) [inline test/augmentation/test_augmentation_mix.py:423]: `kornia.testing.assert_close`

File: test/augmentation/test_augmentation_mix.py [hunk 7]
```diff
@@ -387,3 +387,113 @@ def test_p(self, p, same_on_batch, device, dtype):
         input = torch.randn((12, 3, 256, 256), device=device, dtype=dtype)
 
         f(input)
+
+
+class TestRandomTransplantation:
+    @pytest.mark.parametrize("input_3D", [False, True])
```
Comment (johnnv1) [inline test/augmentation/test_augmentation_mix.py:393]: shouldn't we split the 2d and 3d APIs signature? cc @shijianjian

File: test/augmentation/test_augmentation_mix.py [hunk 8]
```diff
@@ -387,3 +387,113 @@ def test_p(self, p, same_on_batch, device, dtype):
         input = torch.randn((12, 3, 256, 256), device=device, dtype=dtype)
 
         f(input)
+
+
+class TestRandomTransplantation:
```
Comment (johnnv1) [inline test/augmentation/test_augmentation_mix.py:392]: Kornia also has a `BaseTester`, which provides `self.assert_close` to be used instead of `torch.allclose`

[Checkout our testing guide](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.md#coding-standards)

**Review by johnnv1 — COMMENTED** [review]
very interesting augmentation !!

File: kornia/augmentation/_2d/mix/transplantation.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+        self.data_keys = [DataKey.get(inp) for inp in data_keys]
+        self._channel_dim = 1
+
+    def apply_non_transform_mask(self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any]) -> Tensor:
+        return input
+
+    def transform_input(self, acceptor: Tensor, donor: Tensor, selection: Tensor) -> Tensor:
+        # Expand selection to the channel dimension
+        selection = selection.unsqueeze(dim=self._channel_dim).expand_as(donor)
+        acceptor[selection] = donor[selection]
```
Comment (shijianjian) [inline kornia/augmentation/_2d/mix/transplantation.py:148]: This is an in-place operation. I think we shall try avoiding it?

File: kornia/augmentation/_2d/mix/transplantation.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+          receive a transplant.
+        p_batch: probability for applying an augmentation to a batch. This param controls the augmentation
+          probabilities batch-wise.
+        data_keys: the input type sequential for applying augmentations.
+          Accepts "input", "mask".
+
+    Note:
+        - This augmentation requires that segmentation masks are available for all images in the batch and that at
+          least some objects in the image are annotated.
+        - This implementation works for arbitrary spatial dimensions including 2D and 3D images.
```
Comment (shijianjian) [inline kornia/augmentation/_2d/mix/transplantation.py:44]: Not sure. IIRC, the base class would check if all the input are BCHW or not. Probably it is better to make it only support 2D for now. Or if you got time, we may have the 3D MixAugmentationBase.

File: test/augmentation/test_augmentation_mix.py (hunk 7 above)
Comment (shijianjian) [reply test/augmentation/test_augmentation_mix.py:393]: We should. Some checkings are different

File: kornia/augmentation/_2d/mix/transplantation.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+        return input
+
+    def transform_input(self, acceptor: Tensor, donor: Tensor, selection: Tensor) -> Tensor:
+        # Expand selection to the channel dimension
+        selection = selection.unsqueeze(dim=self._channel_dim).expand_as(donor)
+        acceptor[selection] = donor[selection]
+        return acceptor
+
+    def transform_mask(self, acceptor: Tensor, donor: Tensor, selection: Tensor) -> Tensor:
+        acceptor[selection] = donor[selection]
```
Comment (shijianjian) [inline kornia/augmentation/_2d/mix/transplantation.py:152]: Same here for inplace ops

File: kornia/augmentation/_2d/mix/transplantation.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+        return acceptor
+
+    def forward(  # type: ignore[override]
+        self,
+        image: Tensor,
+        mask: Tensor,
+        *additional_inputs: Tensor,
```
Comment (shijianjian) [inline kornia/augmentation/_2d/mix/transplantation.py:159]: Emm... This one is kinda a special case of input. We definitely cannot use it for `AugmentationSequential` if so.

Ideally, only image and mask are required for this forward here. Then the sequential processing for `additional_inputs` shall be supported by `AugmentationSequential`.

I am thinking that, removing the additional_inputs here. Then `AugmentationSequential.__init__` detects if we have a `transplantation`, then throws out a warning that `The first IMAGE and MASK key will be used for transplantation augmentation`. Then we make a special case to use both image and mask for the first round. Then everything else shall be as same as other augmentations, right?

File: kornia/augmentation/_2d/mix/transplantation.py (hunk 9 above)
Comment (JanSellner) [reply kornia/augmentation/_2d/mix/transplantation.py:152]: It is but the assignment happens on a cloned tensor. I could move the `clone()` to the `transform_*` functions, if you prefer.

File: kornia/augmentation/_2d/mix/transplantation.py (hunk 10 above)
Comment (JanSellner) [reply kornia/augmentation/_2d/mix/transplantation.py:48]: The augmentation itself is independent of the spatial dimension since it relies only on the segmentation mask. The current version of the implementation is also independent of the spatial dimensions since I overwrite the `forward()` function. I also changed the test to cover at least 2D, 3D and 4D input.

File: test/augmentation/test_augmentation_mix.py (hunk 7 above)
Comment (JanSellner) [reply test/augmentation/test_augmentation_mix.py]: From a user perspective I would prefer to have only one class since the API is the same for higher dimensional input. A separate `RandomTransplantation3D` class would not do anything different under the hood than the 2D version (or 4D, 5D etc.). If I (as a user) would like to support 2D and 3D input, it is more convenient to have the same class for both cases I think.

File: docs/generate_examples.py (hunk 2 above)
Comment (JanSellner) [reply docs/generate_examples.py:239]: Should I leave this as-is for now since you intend to rewrite the whole script anyway? Otherwise I can of course switch to `kornia.io.write_image`.

File: kornia/augmentation/_2d/mix/transplantation.py (hunk 12 above)
Comment (JanSellner) [reply kornia/augmentation/_2d/mix/transplantation.py]: This is true. I removed the additional arguments and used the same API as the base class with only one `input` argument. I am now using the `data_keys` to identify the first mask tensor and that tensor determines the transplantation which is used for all tensors. This is more flexible now :-) 

But I do see the problem with `AugmentationSequential` now. This does indeed not work since per default only one input at a time is passed on to the transformations. I will try to find a solution for that.

File: docs/generate_examples.py (hunk 2 above)
Comment (johnnv1) [reply docs/generate_examples.py:239]: > Should I leave this as-is for now since you intend to rewrite the whole script anyway? 

yes, we can open an issue later to track this :)

File: test/augmentation/test_augmentation_mix.py (hunk 7 above)
Comment (JanSellner) [reply test/augmentation/test_augmentation_mix.py]: With `AugmentationSequential` there are some differences as then it indeed only works for 2D input. For that reason, I also created a `RandomTransplantation3D` class for 3D input when used together with `AugmentationSequential`. If `RandomTransplantation` is used on its own, it still works for arbitrary spatial dimensions. I also added this information to the docs.

File: kornia/augmentation/_2d/mix/transplantation.py [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+
+        Args:
+            *input: All input tensors passed to the augmentation pipeline.
+            data_keys: Associated data key for every input tensor.
+            params: Dictionary of parameters computed so far by the augmentation pipeline (e.g. including the
+                    `batch_prob`).
+            extra_args: Optional dictionary of extra arguments with specific options for different input types.
+
+        Returns: Updated dictionary of parameters with the necessary information to apply the augmentation on all input
+                 tensors separately.
```
Comment (shijianjian) [inline kornia/augmentation/_2d/mix/transplantation.py:206]:
```suggestion
        Returns:
             Updated dictionary of parameters with the necessary information to apply the augmentation on all input
             tensors separately.
```

File: kornia/augmentation/_3d/mix/transplantation.py [hunk 14]
```diff
@@ -0,0 +1,14 @@
+from kornia.augmentation._2d.mix.transplantation import RandomTransplantation
+from kornia.augmentation._3d.base import AugmentationBase3D
+
+__all__ = ["RandomTransplantation3D"]
+
+
+class RandomTransplantation3D(RandomTransplantation, AugmentationBase3D):  # type: ignore
+    """RandomTransplantation3D augmentation.
+
```
Comment (shijianjian) [inline kornia/augmentation/_3d/mix/transplantation.py:9]: BTW, do you have any 3D images by hand that we can show here?

File: kornia/augmentation/container/ops.py [hunk 15]
```diff
@@ -212,6 +222,9 @@ def transform(cls, input: Tensor, module: Module, param: ParamItem, extra_args:
                 "The support for 3d mask operations are not yet supported. You are welcome to file a PR in our repo."
             )
 
+        elif isinstance(module, K.MixAugmentationBaseV2):
```
Comment (shijianjian) [inline kornia/augmentation/container/ops.py:225]: Probably like below? 
```suggestion
        elif isinstance(module, K.RandomTransplantation):
```

**Review by shijianjian — APPROVED** [review]
LGTM

File: kornia/augmentation/_3d/mix/transplantation.py (hunk 14 above)
Comment (JanSellner) [reply kornia/augmentation/_3d/mix/transplantation.py:9]: Unfortunately not because I never worked with 3D images so far. Some MRI or CT images of a 3D segmentation task would probably be good examples but the visualization is more difficult. But this is better suited for a separate PR IMHO.
